Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp

```diff
@@ -0,0 +1,106 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+
+#include "UserExperienceComponent.h"
+
+#include "Net/UnrealNetwork.h"
+#include "SpatialNetDriver.h"
+#include "Utils/SpatialMetrics.h"
+
+DEFINE_LOG_CATEGORY(LogUserExperienceComponent);
+
+// Sets default values for this component's properties
+UUserExperienceComponent::UUserExperienceComponent()
+{
+	// Set this component to be initialized when the game starts, and to be ticked every frame.  You can turn these features
+	// off to improve performance if you don't need them.
+	PrimaryComponentTick.bCanEverTick = true;
+	bWantsInitializeComponent = true;
+	bAutoActivate = true;
+}
+
+void UUserExperienceComponent::InitializeComponent()
+{
+	UActorComponent::InitializeComponent();
+	ElapsedTime = 0.0f;
+	RequestKey = 0;
+	SetIsReplicated(true);
+}
+
+void UUserExperienceComponent::StartRoundtrip()
+{
+	if (OpenRPCs.Contains(RequestKey))
+	{
+		EndRoundtrip(RequestKey);
+	}
+	int32 NewRPC = RequestKey++;
+	OpenRPCs.Add(NewRPC, ElapsedTime);
+	ServerRTT(NewRPC);
+}
+
+void UUserExperienceComponent::EndRoundtrip(int32 Key)
+{
+	if (float* StartTime = OpenRPCs.Find(Key))
+	{
+		float Diff = ElapsedTime - *StartTime;
+		RoundTripTime.Push(Diff);
+		if (RoundTripTime.Num() > NumWindowSamples)
+		{
+			RoundTripTime.RemoveAt(0);
+		}
+		OpenRPCs.Remove(Key);
+	}
+}
+
+void UUserExperienceComponent::OnRep_ClientTime(float OldValue)
```

Comment:
Oh man - are we sure `OldValue` works in spatial...

Reply:
Do we need to check against the first replication?

Reply:
+ Ignore the first update